migrations/migration.py

```diff
@@ -0,0 +1,178 @@
+            migrations_dir = os.path.join(os.path.dirname(__file__), "versions")
+        self.migrations_dir = migrations_dir
+        self.migration_sequence = self.build_migration_sequence()
+
+    def build_migration_sequence(self) -> dict:
+        """Build the reference sequence dictionary for migration scripts."""
+        migration_sequence = {}
+        migration_files = self.get_migration_files()
+
+        # Dictionary to keep track of nodes visited during traversal
+        visited = set()
+        # Set to keep track of nodes currently being traversed in the DFS
+        traversing = set()
+
+        def dfs(node):
+            if node in traversing:
+                error_message = f"Cycle detected in migration sequence for {node}"
+                audit_entry(
+                    error_message,
+                    extra={"node": node},
+                    level='error'
+                )
+
+                raise ValueError(error_message)
+            if node in visited:
+                return
+            visited.add(node)
+            traversing.add(node)
+            if migration_sequence.get(node):
+                dfs(migration_sequence[node])
+            traversing.remove(node)
+
+        for filename in migration_files:
+            curr_migration = filename[:-3]
+            prev_migration = self.get_previous_migration_id(filename)
+            migration_sequence[curr_migration] = prev_migration
+
+        # Perform DFS from each node
+        for node in migration_sequence:
+            if node not in visited:
+                dfs(node)
+
+        return migration_sequence
+
+    def get_migration_files(self) -> List[str]:
+        """Retrieve the list of migration files."""
+        migration_files = os.listdir(self.migrations_dir)
+        return migration_files
+
+    def get_previous_migration_id(self, filename: str) -> str:
+        """Retrieve the id of the previously ran (downstream) migration script."""
+        prev_migration_id = None
+        with open(os.path.join(self.migrations_dir, filename), "r") as migration_file:
+            for line in migration_file:
+                match = re.match(r"# Previous version: (.+)", line)
+                if match and match.group(1) != 'None':
+                    prev_migration_id = match.group(1)
+                    break
+        return prev_migration_id
+
+    def generate_migration_script(self, migration_name: str):
+        """Generate a new migration script."""
+        current_id = str(self.get_latest_applied_migration_from_fhir())
+        new_id = str(uuid.uuid4())  # Random string as the migration identifier
+        migration_filename = f"{new_id}.py"
+        migration_path = os.path.join(self.migrations_dir, migration_filename)
+        with open(migration_path, "w") as migration_file:
+            migration_file.write(f"# Migration script generated for {migration_name}\n")
+            migration_file.write(f"# Current version: {new_id}\n")
```

Comment:
it would probably be easier to us variables rather than comments for `revision` and `down_revision`, then you can let python do the parsing via `import`.

Reply:
Should I use SourceFileLoader or is there a better practice?